### PR TITLE
[MI-2243] Return error in JSON format from APIs

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -49,13 +50,31 @@ func (p *Plugin) initializeAPI() *mux.Router {
...
+	w.Header().Set("Content-Type", "application/json")
+	errorBytes, err := json.Marshal(apiErr)
+	if err != nil {
+		p.API.LogError("Failed to marshal API error", "error", err.Error())
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(apiErr.StatusCode)
+
```
Comment (ayusht2810) [inline server/plugin/api.go:63]:
```suggestion
```

File: server/plugin/api.go [hunk 2]
```diff
@@ -49,13 +50,31 @@ func (p *Plugin) initializeAPI() *mux.Router {
...
+}
+
 func (p *Plugin) checkAuthBySecret(handleFunc http.HandlerFunc) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		// Replace all occurrences of " " with "+" in WebhookSecret.
 		webhookSecret := strings.ReplaceAll(r.FormValue(SecretParam), " ", "+")
 		if status, err := verifyHTTPSecret(p.getConfiguration().WebhookSecret, webhookSecret); err != nil {
 			p.API.LogError("Invalid secret", "Error", err.Error())
-			http.Error(w, fmt.Sprintf("Invalid Secret. Error: %s", err.Error()), status)
+			p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: status, Message: fmt.Sprintf("Invalid Secret. Error: %s", err.Error())})
```
Comment (ayusht2810) [inline server/plugin/api.go:77]:
```suggestion
			p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: statusCode, Message: fmt.Sprintf("Invalid Secret. Error: %s", err.Error())})
```

File: server/plugin/api.go [hunk 3]
```diff
@@ -127,14 +146,14 @@ func (p *Plugin) handleFileAttachments(w http.ResponseWriter, r *http.Request) {
 	data, appErr := p.API.GetFile(fileInfo.ID)
 	if appErr != nil {
 		p.API.LogError("Couldn't get file data. FileID: %s", fileInfo.ID)
-		http.Error(w, "Couldn't get file data.", http.StatusInternalServerError)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Couldn't get file data."})
```
Comment (ayusht2810) [inline server/plugin/api.go:149]:
```suggestion
		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Couldn't get the file data."})
```

File: server/plugin/api.go [hunk 4]
```diff
@@ -127,14 +146,14 @@ func (p *Plugin) handleFileAttachments(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 
 	w.Header().Set("Content-Type", http.DetectContentType(data))
 	if _, err = w.Write(data); err != nil {
 		p.API.LogError("Error occurred writing the file content in response. Error: %s", err.Error())
-		http.Error(w, "Error occurred writing the file content in response.", http.StatusInternalServerError)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Error occurred writing the file content in response."})
```
Comment (ayusht2810) [inline server/plugin/api.go:156]:
```suggestion
		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Error occurred while writing the file content in response."})
```

File: server/plugin/api.go [hunk 5]
```diff
@@ -311,7 +330,7 @@ func (p *Plugin) handleDateTimeSelectionDialog(w http.ResponseWriter, r *http.Re
 	client := p.MakeClient(r.Context(), token)
 	if err := client.OpenDialogRequest(&requestBody); err != nil {
 		p.API.LogError("Error opening date-time selction dialog.", "Error", err.Error())
-		http.Error(w, "Error opening date-time selection dialog.", http.StatusInternalServerError)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Error opening date-time selection dialog."})
```
Comment (ayusht2810) [inline server/plugin/api.go:333]:
```suggestion
		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusInternalServerError, Message: "Error in opening date-time selection dialog."})
```

File: server/plugin/api.go [hunk 6]
```diff
@@ -257,7 +276,7 @@ func (p *Plugin) handleDateTimeSelectionDialog(w http.ResponseWriter, r *http.Re
 	postActionIntegrationRequest := &model.PostActionIntegrationRequest{}
 	if err := decoder.Decode(&postActionIntegrationRequest); err != nil {
 		p.API.LogError("Error decoding PostActionIntegrationRequest.", "Error", err.Error())
-		http.Error(w, "Error decoding PostActionIntegrationRequest.", http.StatusBadRequest)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusBadRequest, Message: "Error decoding PostActionIntegrationRequest."})
```
Comment (ayusht2810) [inline server/plugin/api.go:279]:
```suggestion
		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusBadRequest, Message: "Error in decoding PostActionIntegrationRequest."})
```

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -954,6 +954,30 @@ func Test_handleDateTimeSelectionDialog(t *testing.T) {
 			},
 			userID: "mock-userID",
 		},
+		"Error opening data/time selection dialog": {
```
Comment (ayusht2810) [inline server/plugin/api_test.go:957]:
```suggestion
		"Error in opening data/time selection dialog": {
```

File: server/plugin/oauth.go [hunk 8]
```diff
@@ -26,25 +27,25 @@ func (p *Plugin) httpOAuth2Connect(w http.ResponseWriter, r *http.Request) {
 func (p *Plugin) httpOAuth2Complete(w http.ResponseWriter, r *http.Request) {
 	code := r.URL.Query().Get("code")
 	if code == "" {
-		http.Error(w, "missing authorization code", http.StatusBadRequest)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusBadRequest, Message: "missing authorization code"})
```
Comment (ayusht2810) [inline server/plugin/oauth.go:30]: we have used error messages starting with the capital letters above. check everywhere

File: server/plugin/oauth.go [hunk 9]
```diff
@@ -26,25 +27,25 @@ func (p *Plugin) httpOAuth2Connect(w http.ResponseWriter, r *http.Request) {
...
 	if state == "" {
-		http.Error(w, "missing authorization state", http.StatusBadRequest)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusBadRequest, Message: "missing authorization state"})
 		return
 	}
 
 	mattermostUserID := r.Header.Get(HeaderMattermostUserID)
 	if mattermostUserID == "" {
-		http.Error(w, "not authorized", http.StatusUnauthorized)
+		p.handleAPIError(w, &serializer.APIErrorResponse{StatusCode: http.StatusUnauthorized, Message: "not authorized"})
```
Comment (ayusht2810) [inline server/plugin/oauth.go:42]: `not authorized` is used at various places. we can make constant for this

File: server/plugin/api.go [hunk 10]
```diff
@@ -49,13 +50,31 @@ func (p *Plugin) initializeAPI() *mux.Router {
 	return r
 }
 
+func (p *Plugin) handleAPIError(w http.ResponseWriter, apiErr *serializer.APIErrorResponse) {
+	w.Header().Set("Content-Type", "application/json")
+	errorBytes, err := json.Marshal(apiErr)
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/plugin/api.go:55]: Took this from `servicenow-classic` plugin. But I think displaying the status code as well is fine.

@manojmalik20 - Your thoughts?

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go [hunk 11]
```diff
@@ -49,13 +50,30 @@ func (p *Plugin) initializeAPI() *mux.Router {
...
+	if err != nil {
+		p.API.LogError("Failed to marshal API error", "error", err.Error())
```
Comment (manojmalik20) [inline server/plugin/api.go:57]:
```suggestion
		p.API.LogError("Failed to marshal API error", "Error", err.Error())
```

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**